aesara/ifelse.py

```diff
@@ -158,86 +150,137 @@ def infer_shape(self, fgraph, node, inputs_shapes):
 
         return out_shapes
 
-    def make_node(self, c, *args):
-        if len(args) != 2 * self.n_outs:
+    def make_node(self, condition: "TensorLike", *true_false_branches: Any):
+        if len(true_false_branches) != 2 * self.n_outs:
             raise ValueError(
-                f"Wrong number of arguments to make_node: expected "
-                f"{int(2 * self.n_outs)}, got {len(args)}"
+                f"Wrong number of arguments: expected "
+                f"{int(2 * self.n_outs)}, got {len(true_false_branches)}"
             )
-        c = at.basic.as_tensor_variable(c)
-        nw_args = []
-        for x in args:
-            if isinstance(x, Variable):
-                nw_args.append(x)
-            else:
-                nw_args.append(at.as_tensor_variable(x))
-        args = nw_args
-        aes = args[: self.n_outs]
-        fs = args[self.n_outs :]
-
-        for t, f in zip(aes, fs):
-            # TODO: Attempt to convert types so that they match?
-            # new_f = t.type.filter_variable(f)
-
-            if not t.type.is_super(f.type):
-                raise TypeError(
-                    "IfElse requires compatible types for true and false return values: "
-                    f"true_branch={t.type}, false_branch={f.type}"
+
+        condition = at.basic.as_tensor_variable(condition)
+
+        if condition.type.ndim > 0:
+            raise TypeError("The condition argument must be a truthy scalar value")
+
+        inputs_true_branch = true_false_branches[: self.n_outs]
+        inputs_false_branch = true_false_branches[self.n_outs :]
+
+        output_vars = []
+        new_inputs_true_branch = []
+        new_inputs_false_branch = []
+        for input_t, input_f in zip(inputs_true_branch, inputs_false_branch):
+
+            if not isinstance(input_t, Variable):
+                input_t = as_symbolic(input_t)
+            if not isinstance(input_f, Variable):
+                input_f = as_symbolic(input_f)
+
+            if isinstance(input_t.type, HasDataType) and isinstance(
+                input_f.type, HasDataType
+            ):
+                # TODO: Be smarter about dtype casting.
+                # up_dtype = aes.upcast(input_t.type.dtype, input_f.type.dtype)
+
+                if input_t.type.dtype != input_f.type.dtype:
+                    raise TypeError(
+                        "IfElse requires compatible types for both branches: got "
+                        f"true_branch={input_t.type}, false_branch={input_f.type}"
+                    )
+
+            if isinstance(input_t.type, HasShape) and isinstance(
+                input_f.type, HasShape
+            ):
+
+                if input_t.type.ndim != input_f.type.ndim:
+                    raise TypeError(
+                        "IfElse requires compatible types for both branches: got "
```

Comment:
The error message could more succinctly say both branches must have the same number of dimensions since that's what's being checked.